### DISCUSSION TITLE
When Pipewire / EasyEffects breaks, it feels like an intelligence test that I am failing:

### DISCUSSION THREAD

**Discussion by AidanofVT** [wwmm/easyeffects#1939] [Q&A]
I'm new to Linux, so before I get frustrated, I try to do my homework and read up on a topic. So, I understand that in Linux, sound is more controllable than in Windows: playback streams must be directed to output devices, and those paths can intersect, divide, and even have additional layers. It's like pipes or wires. Ah, Pipewire!

When I boot up my computer, EasyEffects works pretty sensibly: my playback streams are directed to "EasyEffects Sink", and EasyEffects, though not listed in my OS audio mixer as a playback stream, is directed to the correct output as per its "Pipewire" page. When playback streams change, or I unplug my headphones, all is well: sensible audio pathways are maintained.

--But then something will happen (a Discord call, usually), and it all just-- breaks. I can't really describe it very well, because I can't see a pattern to it. That's why I say it feels like an intelligence test that I'm failing. I can have streams directed to EasyEffects, and EasyEffects directed to an output, but, somewhere along the line, the sound is stopped. Or I can have it, as per the warning, just set to use the default output and "Process all audio streams." In the equalizer effect, the input volume lines can be moving, but not the output lines. Inputs and outputs sometimes stop automatically changing. The "global bypass" button sometimes permits sound, and sometimes doesn't. In the "players" tab of EasyEffects, disabling a player sometimes permits sound, and sometimes doesn't. Sometimes, changing sources/destinations in my OS sound mixer changes whether a player is enabled/disabled in EasyEffects, and vice-versa, often in nonsensical ways.

Is something broken, or am I just not understanding how things work? I keep feeling like there is some combination of check-boxes and drop-down menus that will fix it, but I always end up having to reboot. This issue happens on two different systems, one with Zorin Light and one with SteamOS 3, making me think that the problem is probably me.

Any help/direction/tutorial would be appreciated.

**Comment by wwmm** [comment]
You may be facing a combination of problems. We need to see some logs to understand what may be going on. Kill EasyEffects `easyeffecs -q` and restart it in debug mode `G_MESSAGES_DEBUG=easyeffects easyeffects`. It would help to have this output when things go wrong. As well as the output file of `pw-dot`. It can be viewed with the command `xdot` and will show how filters and devices are linked.

**Reply by AidanofVT** [reply]
Sorry for the newbie-confusion, but I'm having trouble following along: when I open a terminal and enter "easyeffecs -q", I get `easyeffecs: command not found`. I also tried "easyeffects," in case you misspelled it, and "EasyEffects". No dice. I see that there is a "Command Line Options" page on the wiki, but this was no help. Should I be doing these commands in a specific directory? If it helps, I installed EasyEffects through my distro's (Zorin Lite) software store. It says that the source is "dl.flathub.org".

**Reply by wwmm** [reply]
>It says that the source is "dl.flathub.org".

Oh. In this case the Flatpak package was used to do the installation. Take a look at the flatpak command shown at https://github.com/wwmm/easyeffects/wiki/Reporting-Bugs#basic-logs.

**Reply by kjcole** [reply]
It seems quite random whether or not I will get decent audio with the exact same configuration on the exact same applications or websites. Some days, Jitsi, Google Voice, Discord, and Zoom -- whether stand-alone or through a browser -- behave beautifully, other days it's this weird buzzy overlay on whatever's being said or played. 

@wwmm I'm attaching the `EasyEffects` terminal output and the `pw-dump` log. 
[easyeffects.log](https://github.com/user-attachments/files/16241242/easyeffects.log)
[pw-dump.log](https://github.com/user-attachments/files/16241243/pw-dump.log)

As for `coredump` I get:

```
$ flatpak-coredumpctl com.github.wwmm.easyeffects
Traceback (most recent call last):
  File "/usr/bin/flatpak-coredumpctl", line 83, in <module>
    coredumper.run()
  File "/usr/bin/flatpak-coredumpctl", line 44, in run
    subprocess.check_call(["coredumpctl", "dump"] + shlex.split(self.coredumpctl_matches),
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['coredumpctl', 'dump']' returned non-zero exit status 1.
```

**Reply by wwmm** [reply]
> It seems quite random whether or not I will get decent audio with the exact same configuration on the exact same applications or websites. Some days, Jitsi, Google Voice, Discord, and Zoom -- whether stand-alone or through a browser -- behave beautifully, other days it's this weird buzzy overlay on whatever's being said or played.
> 
> @wwmm I'm attaching the `EasyEffects` terminal output and the `pw-dump` log.

@kjcole this isn't the full log. It is important to see what was printed from the beginning. In the file there is only the log from the moment the window is closed on. 

In any case I am confused. What exactly is the problem you are facing? I do not use Discord or Zoom but based on what I have seen in the reports made by users in the last years they definitely do something unusual in their desktop clients. The web version is more reliable because in this case a browser is used.

Maybe there is something messing with the audio routing EasyEffects is trying to do. The command `pw-dot` will be more helpful than pw-dump in this case. Its output file can be viewed with the command xdot and will show each link state.

**Reply by kjcole** [reply]
@wwmm Ah. Thanks for the quick reply. I'm far from an expert here. Of my searches related to the problem, this thread looked the most promising. I'm on Pop! OS 22.04, and, since it seems to like Flatpak, against my preference, I've finally caved. (I may wipe it all and go play with Arch again, but, for the most part, I prefer to use applications rather than fight with them, and Pop! has done a pretty good job as a general rule.)

What I'm trying to do: I am quite often connecting to online gatherings of musicians. I route my mic through both:

- OBS Studio, where I have a virtual camera set up and can do fun things with "synesthesia" like displaying a wave form of the audio at the bottom of the screen, and I'm working up to a "color organ" effect, and
- EasyEffects, where I often add in reverb, and sometimes, other effects

Because I'm using headphones and want to hear the audio with the effects added, the output of EasyEffects goes into whatever meeting software I'm using, and then the output of that source is going into my headphones. (I want to hear what the folks on the other end of the connection are hearing. So, I don't want to have a separate chain where my output from EasyEffects goes directly to my headphones without first going through whatever meeting software I'm using.)

And, I have a separate source that is the output of their mics, going into an EasyEffects sink that generally has no effects turned on, and from there into my headphones. (When things are going bad, I've dropped EasyEffects out of that chain, but it seems to make no difference.)

Tonight, once again, after several random reboots, audio went back to normal. I could flip Reverb on and off, see my wave form, etc.  But for over an hour prior to that, I was getting that buzzing from both my own self-monitoring and from others trying to talk to me.  Today, I sounded normal to them, but that's not always the case. I don't change any settings. I just keep restarting and eventually, something works. (I'm using ``qpwgraph` as my patchbay.) 

It doesn't seem related to bandwidth issues either -- though that's somewhat subjective.

**Reply by wwmm** [reply]
>Tonight, once again, after several random reboots, audio went back to normal. I could flip Reverb on and off, see my wave form, etc. But for over an hour prior to that, I was getting that buzzing from both my own self-monitoring and from others trying to talk to me. Today, I sounded normal to them, but that's not always the case. I don't change any settings. I just keep restarting and eventually, something works.

Maybe in the occasions where things go wrong the actual routing isn't exactly the one you want. OR the latency is getting too low. The command `pw-top` will help with this. Although the `QUANTUM`  and `RATE` column do not tell the whole story about the final latency the ratio `quantum/ratio` is related to it.
